cranelift-faerie/src/backend.rs

```diff
@@ -74,9 +79,164 @@ pub struct FaerieBackend {
     isa: Box<dyn TargetIsa>,
     artifact: faerie::Artifact,
     trap_manifest: Option<FaerieTrapManifest>,
+    frame_sink: Option<FrameSink>,
     libcall_names: Box<dyn Fn(ir::LibCall) -> String>,
 }
 
+struct FrameSink {
+    // we need to retain function names to hand out usize identifiers for FDE addresses,
+    // which are then used to look up function names again for relocations, when `write_address` is
+    // called.
+    fn_names: Vec<String>,
+    table: gimli::write::FrameTable,
+}
+
+impl FrameSink {
+    /// Find a CIE appropriate for the register mapper and initial state provided. This will
+    /// construct a CIE and rely on `gimli` to return an id for an appropriate existing CIE, if
+    /// one exists.
+    ///
+    /// This function also returns a `CFIEncoder` already initialized to the state matching the
+    /// initial CFI instructions for this CIE, ready for use to encode an FDE.
+    pub fn cie_for<'a>(
+        &mut self,
+        initial_state: &[ir::FrameLayoutChange],
+        reg_mapper: &'a DwarfRegMapper,
+    ) -> (gimli::write::CieId, CFIEncoder<'a>) {
+        let mut cie = CommonInformationEntry::new(
+            gimli::Encoding {
+                format: gimli::Format::Dwarf32,
+                version: 1,
+                address_size: 4,
```

Comment:
This probably should be 8. Not sure if it will be used for the data you are writing though.

Reply:
It should only be 8 on x86_64.

Reply:
I had definitely misunderstood the point of the `address_size` field here to be "size of an encoded pointer", rather than "size of a pointer for the target ISA". I'll make this pick the right size for the target.

4 and the asserts later about address size/pointer encoding, are a relic of trying to match CIE bytes I'd known work, in tracking down a crash in `libgcc`. In reality, I eventually figured out the issue I'd had is that `libgcc` doesn't handle escaped length fields in 64-bit DWARF, so I just tried `absptr` encodings and it still works fine! :tada:

I ought to add a note about why this specifically uses `Dwarf32` regardless of architecture - in other places I at least wanted to at least comment about architecture-specific magic values, so `format` deserves a `Do not change this!` mention.

Thinking through my confusion with the meaning of this and the `write_eh_pointer` argument, it seems like `write_eh_pointer` ought to not take `address_size` as a parameter. Since it only matters for absolute pointer encodings, I think it could work better to provide `address_size` as a field of an `absptr` variant for `DwEhPe`? Then it would be clear just from the function signature that the implementer is to write an address matching the encoding, and that `gimli` is not making assumptions about the number of bytes that would be written via `address_size`. If that sounds interesting, I can sketch out a PR to gimli to try that.

Reply:
The `Encoding::address_size` field is used for all DWARF, where the normal format of an address is an unencoded pointer. `.eh_frame` is different from DWARF in that it encodes pointers, but even then it can't be the encoded pointer size because it is possible to use multiple encodings of different sizes.

`DwEhPe` is used to represent the constants, so it can't be changed. The `size` parameter to `write_eh_pointer` is documented as only being used for absolute pointers. I'm not sure we can do better than that. Maybe renaming the parameter to `absolute_address_size` would be worthwhile.

We may want to add a check in gimli to ensure you don't write `.eh_frame` with `Dwarf64`, since the LSB requires 32-bit length fields.